### ISSUE TITLE
Part placement ignored when rendering bodies.

### ISSUE THREAD

**Issue by Brutha12** [FreeCAD/FreeCAD-render#123]
Hi,

I found that the render workbench does not take into account the placement of parts in which bodies are placed when doing the rendering. If the part itself is added to the render workbench instead of the bodies, it works ok for placement - BUT this then prevents the use of multiple materials. This is important in e.g. the case of a wheel - you would have a part representing the wheel consisting of a rubber tyre, a steel hub and an aluminium hub cap for example.

Please see attached file for an example.

Thanks!

Brutha

Render with part added to workbench (correct positioning):
<img width="414" alt="Screenshot 2021-02-21 at 08 20 29" src="https://user-images.githubusercontent.com/66247378/108619920-0cf39e80-7420-11eb-9f21-9826f8df343c.png">

Render with individual bodies added:
<img width="412" alt="Screenshot 2021-02-21 at 08 22 23" src="https://user-images.githubusercontent.com/66247378/108619928-1f6dd800-7420-11eb-88a3-00e8d7a78451.png">

[Render_part_bug.FCStd.zip](https://github.com/FreeCAD/FreeCAD-render/files/6017382/Render_part_bug.FCStd.zip)

**Comment by luzpaz** [comment]
Please append you full [About](https://wiki.freecadweb.org/Aboutl) info to the ticket + please make sure you're using the latest Render WB version (go to the Addon Manager and press install/update button to do so. Thanks.

**Comment by Brutha12** [comment]
> Please append your full [About](https://wiki.freecadweb.org/Aboutl) info to the ticket + please make sure you're using the latest Render WB version (go to the Addon Manager and press install/update button to do so. Thanks.

OK, render workbench updated to most recent - same issue appearing. Version info below.

Thanks!

Brutha

`OS: macOS 10.15
Word size of OS: 64-bit
Word size of FreeCAD: 64-bit
Version: 0.19.23756 (Git)
Build type: Release
Branch: master
Hash: 9c6e9184930a52b165a0b7274e3a45d1006bfe67
Python version: 3.8.6
Qt version: 5.12.5
Coin version: 4.0.0
OCC version: 7.4.0
Locale: C/Default (C)
`

**Comment by howetuft** [comment]
Hello @Brutha12 ,

Thank you for your feedback!

Well, to me, it is more a misuse than a bug: one should not add Bodies individually. Indeed, Body view is basically not aware of encompassing Part placement (as Body per se is not aware of encompassing Part, it works in the opposite way), and I wouldn't want to write some code to bypass that, as it could be quite unstable.
Another point is that, to me, the following principle should be enforced: one part, one material. Complex multi-material parts should then be achieved by assembling mono-material subparts, preferably using Assembly features, but alternatively embedding Parts in Part, like you did with Part001 and Part.

However, your remark is very judicious in that it points out the absence of a key feature in FreeCAD: to achieve what I described above, _Part should have a serviceable Material property_, which is not the case at the moment.

I'm gonna ask for such an evolution upstream in FreeCAD.

**Comment by Brutha12** [comment]
> Another point is that, to me, the following principle should be enforced: one part, one material. Complex multi-material parts should then be achieved by assembling mono-material subparts, preferably using Assembly features, but alternatively embedding Parts in Part, like you did with Part001 and Part.

Thanks for taking a look! 

You're right: we need some way of assigning materials OUTSIDE the render workbench I think. Since each Part object is also not aware of its parent I assume, only the top level Part object should ever be added to the view.

Assigning materials at Part level means you have to add some logic to decide which material to use in the case of conflicting assignments, and also requires that you encompass each body in a a separate Part.

On the other hand, if you assign material at Body level, then it restricts people to using only the Part Design work bench I suppose.

Cheers

Brutha

**Comment by Brutha12** [comment]
> Since each Part object is also not aware of its parent I assume, only the top level Part object should ever be added to the view.

Although, thinking about it, this brings another problem. 

If I am building my project in one document, and I want to render only one of the sub assemblies, I cannot see where to position my camera, since in the viewport I will see its real location.

**Comment by luzpaz** [comment]
@Brutha12 please also post the forum thread you opened before opening this ticket. That way we can keep it up to date

**Comment by Brutha12** [comment]
> @Brutha12 please also post the forum thread you opened before opening this ticket. That way we can keep it up to date

Thanks - the thread is [https://forum.freecadweb.org/viewtopic.php?f=3&t=55812&sid=6df6e7ff38491029fa36cf034ee647c9](url).

Although - not sure if I understood correctly, shall we add this discussion back to the thread do you think?

**Comment by howetuft** [comment]
> Assigning materials at Part level means you have to add some logic to decide which material to use in the case of conflicting assignments, and also requires that you encompass each body in a a separate Part.

The rule (already enforced with some other objects) will be: the upper-level material, if existing, overrides the lower-level one.
It means that, if we have:

CompoundPart (_No Material_)
   |-- SubPart1 (Material#1)
   |-- SubPart2 (Material#2)

...we shall see SubPart1 in Material1 and SubPart2 in Material2.

And if we have:

CompoundPart (Material#3)
   |-- SubPart1 (Material#1)
   |-- SubPart2 (Material#2)

...we shall see both SubPart1 and SubPart2 in Material3.

In my opinion, it is yet complex enough: that is why I would be reticent to manage also a Material property on Bodies...

> If I am building my project in one document, and I want to render only one of the sub assemblies, I cannot see where to position my camera, since in the viewport I will see its real location.

Sorry, I'm not sure to understand exactly what you mean: could you please provide an example?

**Comment by luzpaz** [comment]
> > @Brutha12 please also post the forum thread you opened before opening this ticket. That way we can keep it up to date
> 
> Thanks - the thread is [https://forum.freecadweb.org/viewtopic.php?f=3&t=55812&sid=6df6e7ff38491029fa36cf034ee647c9](url).
> 
> Although - not sure if I understood correctly, shall we add this discussion back to the thread do you think?

No just post the conclusion of this discussion here and mark it [Solved] when it is. :wink:

**Comment by Brutha12** [comment]
> Sorry, I'm not sure to understand exactly what you mean: could you please provide an example?

Sure, let's say I have the following structure, and each of the Parts have some placement information:

Main part
-> Sub Part 1
-> Sub Part 2
----> Sub Part 2a
----> Sub Part 2b
-> Sub Part 3

If I want a render of the whole project, I can add "Main part" to the render project. The rendered view will correspond to what I see in the viewport.

However, if I add instead only"Sub Part 2b" to the render project, then the placement information in "Sub Part 2" and "Main part" are ignored (at least, from my testing!). So, the view from the camera no longer corresponds to what I see in the viewport.

**Comment by Brutha12** [comment]
Thinking out loud, I am wondering if it might be a good approach to do this:

- By default, you include *all* objects in the project in the render, using all placement information in parent Parts - in the same way as the viewport.

- Allow the user to add objects to the render project and *override* the behaviour. For example, adding material information at whatever level, and specifying various render options - for example "Include this object and children in the render" or "Exclude this object and children from the render".

Just a thought though!

**Comment by howetuft** [comment]
Posted a PR to FreeCAD to make App::Part.Material usable (the same way as Arch objects): wait and see...

@Brutha12 For your proposition: from user point of view, it may be more convenient for sure. But there are also a few objections to overcome:
- View object is a convenient place to store some specific information, and especially what is called "View result", which is a cached result from translation to renderer's SDL. In a word, it is essential to 'Delayed Build' feature, so if we get rid of systematic View usage, we have to forego this feature. As it was a specific aim of the original designers of the workbench, it needs a discussion...
- There are a lot of unrenderable objects in FreeCAD (see workbenches: BIM, Draft, Materials - and even Render WB's objects...). So beforehand you should precise how you decide an object will be rendered or not.

**Comment by howetuft** [comment]
> However, if I add instead only"Sub Part 2b" to the render project, then the placement information in "Sub Part 2" and "Main part" are ignored (at least, from my testing!). So, the view from the camera no longer corresponds to what I see in the viewport.

I think you can use Visible property on each subpart. For instance, with your model, I unchecked Visible for Body003, and I get the following, with standard camera:

![image](https://user-images.githubusercontent.com/34142192/108635129-d6e50780-747d-11eb-8c7d-d15f7027d225.png)

**Comment by howetuft** [comment]
Hello @Brutha12,

I finally implemented the feature you suggested, making it possible to add a Material to PartDesign Body!
There is now a "Material Applier" in the toolbar, which one can use to apply a given Material to whatever he/she wants to render, including Body (see documentation [here](https://github.com/FreeCAD/FreeCAD-render/blob/master/docs/Materials.md)). This way, one is not required to use the ugly Material property of the View.

One exception, however: at this time, it's not possible to assign a Material to a Part, due to some particulars in this object (I posted a PR to FreeCAD to circumvent it).

As an illustration, I reworked your example with 4 materials:
![CyclesProject001p4ugf1s4](https://user-images.githubusercontent.com/34142192/111026580-2cf8fb00-83eb-11eb-9a51-3047bb0ba30d.png)


Your feedback is welcome!
